### PR TITLE
ISPN-8756 Single port protocol autodetection

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-8756

Recycles https://github.com/infinispan/infinispan/pull/6133 and simplifies it by adding a simple Hot Rod header magic detector which automatically switches protocol without requiring an explicit HTTP Upgrade header.

### CODE REVIEW

File: documentation/src/main/asciidoc/server_guide/server_guide.asciidoc [hunk 1]
```diff
@@ -6,3 +6,122 @@ The {brandname} community
...
+ConfigurationBuilder builder = new ConfigurationBuilder()
+      .addServers("127.0.0.1:8080");
+
+RemoteCacheManager remoteCacheManager = new RemoteCacheManager(builder.build());
+RemoteCache<String, String> cache = remoteCacheManager.getCache("default"");
+
+----
+
+The Hotrod client detects if Single Port has been enabled on the server by checking the port number. If the port is
+specific to REST (ports 80, 8080, 8443 or similar), the Single Port functionality will be enabled on the client.
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/server_guide/server_guide.asciidoc]: Unfortunately I had not amended that part of the document with the latest changes. No settings are required on the client side and no specific ports must be used.

File: server/router/src/main/java/org/infinispan/server/router/router/impl/singleport/SinglePortChannelInitializer.java [hunk 2]
```diff
@@ -23,11 +24,16 @@
    public SinglePortChannelInitializer(SinglePortEndpointRouter server, NettyTransport transport, RestServer restServer, Map<String, ProtocolServer> upgradeServers) {
       super(server, transport, null, null);
       http11To2UpgradeHandler = new SinglePortUpgradeHandler(server.getConfiguration().ssl().enabled(), restServer, upgradeServers);
+
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/router/src/main/java/org/infinispan/server/router/router/impl/singleport/SinglePortChannelInitializer.java]: Removed

**Comment by tristantarrant** [comment]
I don't want to document this in detail for the legacy server: I want to make this the default configuration for ServerNG.

**Comment by wburns** [comment]
> I don't want to document this in detail for the legacy server: I want to make this the default configuration for ServerNG.

Okay, thanks for clarification.

**Comment by wburns** [comment]
Integrated into master, thanks @tristantarrant !

File: server/router/src/main/java/org/infinispan/server/router/router/impl/singleport/HotRodPingDetector.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.infinispan.server.hotrod.HotRodServer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+public class HotRodPingDetector extends ByteToMessageDecoder {
```
Comment (slaskawi) [inline server/router/src/main/java/org/infinispan/server/router/router/impl/singleport/HotRodPingDetector.java]: Very clever, I like it!

**Comment by slaskawi** [comment]
Great work @tristantarrant !

And thank you for putting effort in amending my commits!!!

**Pull request merged**